### DISCUSSION TITLE
FastBufferedWriter.flush() does not flush the underlying Writer

### DISCUSSION THREAD

**Discussion by mreid-exiger** [osiegmar/FastCSV#122] [General]
**Describe the bug**

When using the internally buffered writer, there is no way to flush the underlying Writer. It initially appears this should work, as the when supplying your own Writer, the `flushWriter` flag is true, which causes the `Writer.flush()` to be called when a record is completed; but for FastBufferedWriter, only the internal buffer gets flushed, the underlying Writer is never actually flushed:

```
        @Override
        public void flush() throws IOException {
            writer.write(buf, 0, pos);
            pos = 0;
        }
```

Arguably you don't want to flush the underlying buffer after each record, but if that's the case, it certainly would be helpful to allow the caller to control when the writer is flushed. Presently the only way to flush the underlying Writer is to close it.

**To Reproduce**
JUnit test to reproduce the behavior:
```
@Test
void testFlush() throws Exception {
  CsvWriter csvWriter = CsvWriter.builder().build(new OutputStreamWriter(System.out, "UTF-8"));
  csvWriter.writeRecord("one", "two", "three");
  // Nothing happens until CsvWriter.close() is called
}
```

**Additional context**

I suppose its atypical to use System.out as the target of CsvWriter, but I do find that useful, particularly when utilities are being developed I prefer to look at output on the console rather than writing into a file.

**Comment by osiegmar** [comment]
I don't see where this is a bug.

If you want to use your own writer for whatever reason, just call `flush()` on that one if you want it to flush. The CsvWriter already sent all the data to it.

```java
OutputStreamWriter writer = new OutputStreamWriter(System.out, "UTF-8");
CsvWriter csvWriter = CsvWriter.builder().build(writer);
csvWriter.writeRecord("one", "two", "three");
writer.flush();
```

**Comment by mreid-exiger** [comment]
Certainly this can be worked around with relatively little effort. I reported as a bug as it goes against the typical pattern for wrapping Writers. For example, all wrappers in standard library flush the wrapped writer/stream:

- java.io.BufferedWriter
- java.io.FilterWriter
- java.io.OutputStreamWriter (flushes StreamEncoder)
- java.io.PrinterWriter
- etc.

This remains true for any other I/O wrappers I can remember dealing with in other libraries.

The behavoiur here caught me by surprise as it goes against the rest of the Java ecosystem.

**Reply by osiegmar** [reply]
All the writer examples you mentioned inherit from `java.io.Writer`! They are designed to wrap other writers, maintaining the same interface while extending functionality.

In contrast, the `CsvWriter` is not a writer in the sense of a `java.io.Writer` inheritor. It does not implement methods like `write(char[], int, int)`, `append(char)`, or many others either. The `java.io.Writer` class implements the `Appendable`, `Closeable`, and `Flushable` interfaces, whereas I chose to implement only `Closeable` for the `CsvWriter`.

Nevertheless, thank you for the feedback. I will consider implementing `Flushable` in the future and will keep this issue open for now.

**Comment by osiegmar** [comment]
I've implemented multiple functionality with 0f6bf93.

## Implement `Flushable` interface for `CsvWriter` to allow flushing the underlying writer

```java
CsvWriter csv = CsvWriter.builder().build(new OutputStreamWriter(System.out, "UTF-8"));
csvWriter.writeRecord("one", "two", "three");

// This would be possible now:
csv.flush();
```

## Implement `autoFlush` option for `CsvWriter` to automatically flush the writer after writing a record

```java
CsvWriter csv = CsvWriter.builder()
    .autoFlush(true) // Enable autoFlush
    .build(new OutputStreamWriter(System.out, "UTF-8"));

// Since autoFlush is enabled, the following would automatically flush the writer
csv.writeRecord("one", "two", "three");
```

## Implement `toConsole` method for `CsvWriter` to write records to the console

A more convenient way to write records to the console:

```java
CsvWriter csv = CsvWriter.builder().toConsole();

// Since toConsole is used, the following would write the record directly to the console
csv.writeRecord("one", "two", "three");
```

A release will follow soon.

**Reply by mreid-exiger** [reply]
Very nice, thanks!
